src/fsharp/FSharp.Core/string.fs

```diff
@@ -21,16 +21,52 @@ namespace Microsoft.FSharp.Core
 
         [<CompiledName("Iterate")>]
         let iter (action : (char -> unit)) (str:string) =
-            if not (String.IsNullOrEmpty str) then
-                for i = 0 to str.Length - 1 do
-                    action str.[i] 
+            let len = length str
+
+            if len = 0 then ()
+            else
+                // Unrolling gives a ~25% boost over a regular for loop on an older processor, more on a modern processor
+                let mutable i = 0
+                while i < len - len % 8 do
+                    action str.[i]
+                    action str.[i + 1]
+                    action str.[i + 2]
+                    action str.[i + 3]
+                    action str.[i + 4]
+                    action str.[i + 5]
+                    action str.[i + 6]
+                    action str.[i + 7]
+                    i <- i + 8
+
+                // The remainder
+                while i < len do
+                    action str.[i]
+                    i <- i + 1
 
         [<CompiledName("IterateIndexed")>]
         let iteri action (str:string) =
-            if not (String.IsNullOrEmpty str) then
+            let len = length str
+            if len = 0 then ()
+            else
                 let f = OptimizedClosures.FSharpFunc<_,_,_>.Adapt(action)
-                for i = 0 to str.Length - 1 do
-                    f.Invoke(i, str.[i]) 
+
+                // Unrolling gives a ~25% boost over a regular for loop on an older processor, more on a modern processor
+                let mutable i = 0
+                while i < len - len % 8 do
+                    f.Invoke(i, str.[i])
```

Comment:
Not a good look. And highly likely that the performance improvement is platform dependent.  Raise a bug with the Jit team to do a better job.

Reply:
Same as above.